_posts/pl/2021-12-XX-transactions.md

```diff
@@ -0,0 +1,217 @@
+| `TRANSACTION_REPEATABLE_READ`  | Uniemożliwione | Uniemożliwione         | Dozwolone         |
+| `TRANSACTION_SERIALIZABLE`     | Uniemożliwione | Uniemożliwione         | Uniemożliwione    |
+
+## Punkty zapisu (_ang. save points_)
+Oprócz operacji zatwierdzania transakcji (_COMMIT_) oraz jej wycofywania w przypadku niepowodzenia (_ROLLBACK_),  istotna jest jeszcze jedna funkcja całego mechanizmu – **punkty zapisu**. Są to miejsca, do których nastąpi wycofanie transakcji, w przypadku awarii czy niepowodzenia. Posługując się fragmentem kodu, zastosowanie punktów zapisu, mogłoby wyglądać tak:
+```java
+// ustawienie poziomu izolacji
+connection.setTransactionIsolation(Connection.TRANSACTION_READ_UNCOMMITED);
+
+// instrukcje wykonywane w ramach transakcji – część 1
+// …
+
+// ustawienie punktu zapisu 
+Savepoint savepoint = connection.setSavepoint();
+
+// instrukcje wykonywane w ramach transakcji – część 2 
+// … 
+
+// przywrócenie zmian do stanu z punktu zapisu
+connection.rollback(savePoint);
+```
+
+# Transakcje w Springu
+O ile JDBC pozwala na zarządzanie transakcjami poprzez wykorzystanie _setAutocommit(false)_, o tyle Spring daje kilka sposobów osiągnięcia tego samego w znacznie wygodniejszy sposób.
+
+## Programowalne zarządzanie transakcją
+Pierwsza, ale raczej rzadko wykorzystywana metoda to użycie _TransactionTemplate_ albo _PlatformTransactionManager_. Sposób jej użycia wygląda następująco:
+```java
+@Service
+public class BookService {
+
+	@Autowired
+	private TransactionTemplate template;
+
+	public Long addBook(Book book) {
+		Long id = template.execute(status → {
+			// dodaj książkę do bazy i zwróć wygenerowany identyfikator 
+			// … 
+			return id;
+		});
+	}
+}
+```
+**Zalety**:
+- nie trzeba zaprzątać sobie głowy otwieraniem i zamykaniem połączenia z bazą danych poprzez stosowanie bloku _try-catch-finally_. Zamiast tego wykorzystywany jest mechanizm callback’ów (`Transaction Callbacks`).
+- nie trzeba przechwytywać wyjątków _SQLException_, ponieważ Spring konwertuje je na wyjątki _RuntimeException_.
+- lepsza integracja z ekosystemem Springa. _TransactionTemplate_ wykorzystuje _TransactionManager_ do konfiguracji połączenie z bazą danych. Ich implementacja wiążę się z koniecznością utworzenia odpowiednich beanów, jednak później nie musimy się już martwić o zarządzanie nimi.
+
+## Deklaratywne zarządzanie transakcją
+O ile powyższa metoda jest znacznym usprawnieniem i ułatwieniem w zarządzaniu transakcjami, to  Spring daje nam nieco więcej, a wszystko za sprawą adnotacji **`@Transactional`**. Dzięki niej, każda publiczna metoda, która zostanie oznaczona tą adnotacją, zostanie wykonana wewnątrz transakcji.
+Mając już wiedzę w jaki sposób wygląda zarządzanie transakcją w JDBC, zobaczmy jak łatwo przekształcić tak, aby wykorzystywał dobrodziejstwo frameworka Spring:
+
+Na początek zobaczmy jak wyglądałoby dodanie nowej książki do bazy przy użyciu JDBC:
```

Comment:
ten kod juz raz pojawił się we wpisie, może lepiej sie do niego odwołać zamiast przeklejać

Reply:
prawda, dodałem odwołanie